src/common/stack_mmap.c

```diff
@@ -0,0 +1,343 @@
+/**
+ * (C) Copyright 2016-2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * \file
+ *
+ * This file is part of the DAOS server. It implements an alternate way
+ * for ULTs stacks allocation, based on mmap() of MAP_STACK|MAP_GROWSDOWN
+ * regions, in order to allow overrun detection along with automatic growth
+ * possibility.
+ */
+
+#ifdef ULT_MMAP_STACK
+#include <gurt/atomic.h>
+#include <daos/common.h>
+#include <daos/stack_mmap.h>
+#include <errno.h>
+#include <string.h>
+
+/* ABT_key for mmap()'ed ULT stacks */
+ABT_key stack_key;
+
+/* pool of free stacks */
+pthread_mutex_t stack_free_list_lock;
+d_list_t stack_free_list;
+static uint64_t alloced_stacks, free_stacks;
+#define MAX_PERCENT_FREE_STACKS 20
+#define MAX_NUMBER_FREE_STACKS 2000
+
+/* callback to free stack upon ULT exit during stack_key deregister */
+void free_stack(void *arg)
+{
+	mmap_stack_desc_t *desc = (mmap_stack_desc_t *)arg;
+	bool do_munmap = false;
+	uint64_t shadow_alloced_stacks, shadow_free_stacks;
+
+	/* XXX
+	 * We may need to reevaluate stack size since a growth may
+	 * have occurred during previous context life time, if initial
+	 * stack size has overflowed when there was no previous mapping
+	 * in address space to trigger Kernel's stack guard gap
+	 * (stack_guard_gap) ? This for both munmap() or queuing in
+	 * free pool cases.
+	 */
+
+	D_MUTEX_LOCK(&stack_free_list_lock);
+	shadow_alloced_stacks = alloced_stacks;
+	shadow_free_stacks = free_stacks;
+	if (free_stack > MAX_NUMBER_FREE_STACKS &&
+	    free_stacks/alloced_stacks * 100 > MAX_PERCENT_FREE_STACKS) {
+		do_munmap = true;
+		--alloced_stacks;
+	} else {
+		d_list_add_tail(&desc->stack_list, &stack_free_list);
+		++free_stacks;
+	}
+	D_MUTEX_UNLOCK(&stack_free_list_lock);
+	if (do_munmap) {
+		int rc;
+
+		D_DEBUG(DB_MEM, "%p mmap()'ed stack of size %zd is beeing munmap()'ed, with alloced_stacks="DF_U64" and free_stacks="DF_U64"\n",
+			desc->stack, desc->stack_size, shadow_alloced_stacks,
+			shadow_free_stacks);
+		rc = munmap(desc->stack, desc->stack_size);
+		/* XXX
+		 * should we re-queue it on free list instead to leak it ?
+		 */
+		if (rc != 0)
+			D_ERROR("Failed to munmap() %p stack of size %zd : %s\n",
+				desc->stack, desc->stack_size, strerror(errno));
+	} else {
+		D_DEBUG(DB_MEM, "%p mmap()'ed stack of size %zd has been put on stack free list, with alloced_stacks="DF_U64" and free_stacks="DF_U64"\n",
```

Comment:
(style)  line over 100 characters

Reply:
can break up like suggestion above

Reply:
Concerning the line length limit, I thought there was the agreement it does not concern the log/msgs strings ?